### ISSUE TITLE
Get a Progress or ProgressReporting object as a return from generate data

### ISSUE THREAD

**Issue by omiz** [techprimate/TPPDF#155]
Is it possible to change handing the progress from a callback closure passed in the generate function of `PDFGenerator` to be a return value 

for better understanding what I mean I would suggest watching this wwcd session:

Best Practices for Progress Reporting
https://developer.apple.com/videos/play/wwdc2015/232/

**Comment by techprimate-phil** [comment]
Yes, this is a great input! Thank you for that!

Definitely going to take a look at it, as the current progress reporting is pretty bad

**Comment by techprimate-phil** [comment]
Hi there, it's been a while but I hope you are still interested in this feature.
I had to remove the static functions from the `PDFGenerator` (I wanted to do that anyways) but it should now have an instance of progress available.

Feel free to checkout the branch `feature/155-progress-reporting` and let me know about your thoughts

**Comment by omiz** [comment]
I took a quick look at the progress reporting branch and it looks amazing. thank you a lot for this great work.

There is one thing that I would like your advice about:
I noticed that you're using ```init(totalUnitCount unitCount: Int64)``` to initialize the Progress sometimes. I also read from the apple documentation before 

> The initializer is passed the current progress object, if there is one, and the value of the totalUnitCount property is set.

Do you think that might cause a problem while running multiple instances of the builder at the same time?

By the way I personally like to initiate the progress object without any parameters ```Progress()``` which will set the parent progress to nil (a longer version would be ```init(parent parentProgressOrNil: Progress?, userInfo userInfoOrNil: [ProgressUserInfoKey : Any]? = nil)```) and then add children through the method ```func addChild(_ child: Progress, withPendingUnitCount inUnitCount: Int64)```

**Comment by techprimate-phil** [comment]
Thats a good question. I just took another look and there is a static factory method  `Progress.discreteProgress(totalUnitCount: Int64)` where the parent is set to `null`.

> Creates and returns an NSProgress instance with the specified totalUnitCount that is not part of any existing progress tree. The instance is initialized using init(parent:userInfo:) with the parent set to nil.

I just changed all initialiser to use `discreteProgress` and it seems to work fine.

About the threading issue:
I guess if each TPPDF `PDFGenerator` runs in it own thread it should not collide. As the framework does not actually provide any async methods for now, this is a responsibility of the user
